modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java

```diff
@@ -2027,25 +2035,27 @@ private Result invokeDown(final Invoke x, final long pageId, final long backId,
                             continue;
                         }
 
-                        // Unfinished Put does insertion on the same level.
-                        if (x.isPut())
-                            continue;
-
-                        assert x.isRemove(); // Guarded by isFinished.
+                        assert x.op != null; // Guarded by isFinished.
 
-                        res = x.finishOrLockTail(pageId, page, backId, fwdId, lvl);
+                        res = x.op.finishOrLockTail(pageId, page, backId, fwdId, lvl);
 
                         return res;
 
                     case NOT_FOUND:
                         if (lvl == 0)
                             x.invokeClosure();
 
+                        // Level must be equal to bottom level. This is the place when we would insert values into
+                        // parent nodes during splits.
+                        assert lvl == (x.isPut() ? ((Put)x.op).btmLvl : 0) : "NOT_FOUND on the wrong level: lvl=" + lvl;
```

Comment:
How about adding more information to the assert? Would `btmLvl` be useful in case this invariant is broken?

Reply:
This was more about documenting the code, but sure, I'll add bottom level to the output.